File: src/claw_mem/decay/tiered_decay.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
TieredDecayEngine (F2 · v4.7.0)

Three-tier storage decay: HOT (current session), WARM (recently accessed),
COLD (long-term). Evicts based on composite score of recency, frequency,
and importance.
"""

import time
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from ..memory_manager import MemoryManager
    from ..llm_provider import LLMProvider
    from .functions import DecayConfig


_IMPORTANCE_PROMPT = (
    "Rate the importance of this memory on a scale of 0.0 to 1.0 "
    "(0=trivial, 0.5=useful, 0.8=very important, 1.0=critical). "
    "Return only the number.\n\nMemory: {content}"
)


class TierLevel(Enum):
    HOT = "hot"
    WARM = "warm"
    COLD = "cold"


def _now_iso() -> str:
    return datetime.now().isoformat()


class TieredDecayEngine:
    """Three-tier storage decay engine for memory lifecycle management.

    Memories flow: HOT → WARM → COLD → EVICTED.
    Access promotes memories upward; time pushes them downward.
    Composite scoring decides eviction order.
    """

    def __init__(
        self,
        manager: "MemoryManager",
        config: "DecayConfig" = None,
        llm_provider: "LLMProvider" = None,
        hot_ttl: int = 3600,
        warm_ttl_days: int = 7,
        cold_ttl_days: int = 30,
        max_hot: int = 100,
        max_warm: int = 500,
        max_cold: int = 2000,
    ):
        self.manager = manager
        self.config = config
        self.llm_provider = llm_provider

        self.hot_ttl = hot_ttl          # seconds
        self.warm_ttl = warm_ttl_days * 86400
        self.cold_ttl = cold_ttl_days * 86400
        self.max_hot = max_hot
        self.max_warm = max_warm
        self.max_cold = max_cold

        # Track access timestamps for recency scoring
        self._access_log: Dict[str, List[float]] = {}
        self._importance_cache: Dict[str, float] = {}
        self._last_cycle: float = 0.0

    # ── classification ─────────────────────────────────────────────────

    def classify(self, memory: Dict[str, Any]) -> TierLevel:
        """Classify a single memory into a tier based on its age and metadata."""
        meta = memory.get("metadata", {})

        # Deprecated memories go directly to COLD for immediate eviction
        if meta.get("deprecated") in ("true", "True", "1"):
            return TierLevel.COLD

        # Use creation timestamp for age calculation
        created_at = memory.get("created_at") or memory.get("timestamp") or _now_iso()
        try:
            created_dt = datetime.fromisoformat(created_at.replace("Z", "+00:00"))
        except (ValueError, TypeError):
            created_dt = datetime.now()

        age_seconds = (datetime.now() - created_dt.replace(tzinfo=None)).total_seconds()

        if age_seconds <= self.hot_ttl:
            return TierLevel.HOT
        elif age_seconds <= self.warm_ttl:
            return TierLevel.WARM
        else:
            return TierLevel.COLD

    def promote(self, memory_id: str) -> Optional[TierLevel]:
        """Record an access, potentially promoting the memory tier.

        Promotion rules:
          COLD → WARM if recently accessed (resets TTL)
          WARM → HOT if accessed within hot TTL
          HOT stays HOT
        """
        if not memory_id:
            return None

        now = time.time()
        if memory_id not in self._access_log:
            self._access_log[memory_id] = []
        self._access_log[memory_id].append(now)

        # Look up current memory
        all_m = self.manager.semantic.get_all()
        mem = next((m for m in all_m if m.get("id") == memory_id), None)
        if mem is None:
            return None

        current = self.classify(mem)
        if current in (TierLevel.HOT, TierLevel.WARM):
            return current
        # COLD → WARM promotion: record access and return
        freq = self._access_frequency(memory_id)
        return TierLevel.WARM if freq >= 2 else TierLevel.COLD

    # ── importance scoring ─────────────────────────────────────────────

    def get_importance(self, memory: Dict[str, Any]) -> float:
        """Score memory importance (0.0–1.0).

        Uses cached LLM score or rule-based fallback.
        """
        mid = memory.get("id", "")
        if mid and mid in self._importance_cache:
            return self._importance_cache[mid]

        # Try LLM scoring
        if self.llm_provider:
            try:
                content = memory.get("content", "")
                score_text = self.llm_provider.generate(
                    _IMPORTANCE_PROMPT.format(content=content),
                    max_tokens=16,
                )
                score = self._parse_score(score_text)
                if score is not None:
                    if mid:
                        self._importance_cache[mid] = score
                    return score
            except Exception:
                pass

        # Rule-based fallback
        score = self._rule_importance(memory)
        if mid:
            self._importance_cache[mid] = score
        return score

    @staticmethod
    def _parse_score(text: str) -> Optional[float]:
        if not text:
            return None
        try:
            score = float(text.strip())
            return max(0.0, min(1.0, score))
        except ValueError:
            return None

    @staticmethod
    def _rule_importance(memory: Dict[str, Any]) -> float:
        """Rule-based importance fallback when LLM is unavailable."""
        content = memory.get("content", "")
        tags = memory.get("tags", [])
        meta = memory.get("metadata", {})
        score = 0.3  # baseline

        # Longer content may be more important
        if len(content) > 200:
            score += 0.15
        elif len(content) > 50:
            score += 0.05

        # Presence of tags
        if tags:
            score += 0.05

        # Critical tags
        critical_keywords = ("critical", "important", "essential", "preference", "rule")
        if any(kw in " ".join(tags).lower() for kw in critical_keywords):
            score += 0.2

        # Explicit importance in metadata
        if meta.get("importance"):
            try:
                score = float(meta["importance"])
            except (ValueError, TypeError):
                pass

        return max(0.0, min(1.0, score))

    # ── access metrics ─────────────────────────────────────────────────

    def _access_frequency(self, memory_id: str) -> int:
        timestamps = self._access_log.get(memory_id, [])
        now = time.time()
        # Count accesses within the last 14 days
        cutoff = now - 14 * 86400
        return sum(1 for ts in timestamps if ts >= cutoff)

    def _access_recency(self, memory_id: str) -> float:
        """Seconds since last access (lower = more recent)."""
        timestamps = self._access_log.get(memory_id, [])
        if not timestamps:
            return float("inf")
        return time.time() - max(timestamps)

    # ── eviction ───────────────────────────────────────────────────────

    def _composite_score(self, memory: Dict[str, Any]) -> float:
        """Composite eviction score. Lower score = evict first.

        Formula: norm_recency * 0.4 + norm_freq * 0.3 + importance * 0.3
        """
        mid = memory.get("id", "")
        importance = self.get_importance(memory)

        recency = self._access_recency(mid)
        # Normalize: recently accessed → higher score
        if recency == float("inf") or recency <= 0:
            norm_rec = 0.0
        else:
            ttl_x2 = self.cold_ttl * 2
            norm_rec = max(0.0, 1.0 - recency / ttl_x2)

        freq = self._access_frequency(mid)
        norm_freq = min(1.0, freq / 5.0)  # cap at 5 accesses

        return norm_rec * 0.4 + norm_freq * 0.3 + importance * 0.3

    def evict(self) -> int:
        """Evict low-score memories, preferentially from COLD tier.

        Always evicts deprecated memories first, then applies composite scoring.
        Returns count of evicted memories.
        """
        storage = self.manager.semantic
        all_memories = storage.get_all()

        evicted_ids: List[str] = []
        non_evicted: List[Dict[str, Any]] = []

        # Phase 1: always evict deprecated
        for m in all_memories:
            meta = m.get("metadata", {})
            if meta.get("deprecated") in ("true", "True", "1"):
                evicted_ids.append(m.get("id", ""))
            else:
                non_evicted.append(m)

        # Respect max tier capacities
        # Gather per-tier counts
        tiers: Dict[TierLevel, List[Dict[str, Any]]] = {
            TierLevel.HOT: [],
            TierLevel.WARM: [],
            TierLevel.COLD: [],
        }
        for m in non_evicted:
            tier = self.classify(m)
            tiers[tier].append(m)

        # Evict from COLD tier if over capacity
        max_map = {
            TierLevel.HOT: self.max_hot,
            TierLevel.WARM: self.max_warm,
            TierLevel.COLD: self.max_cold,
        }

        for tier, max_cap in max_map.items():
            if tier == TierLevel.HOT:
                continue  # Don't evict from HOT tier
            tier_mems = tiers[tier]
            overflow = len(tier_mems) - max_cap
            if overflow <= 0:
                continue
            # Score and sort: lowest score first
            scored = [(self._composite_score(m), m) for m in tier_mems]
            scored.sort(key=lambda x: x[0])
            for _, m in scored[:overflow]:
                evicted_ids.append(m.get("id", ""))

        if not evicted_ids:
            return 0

        # Mark evicted memories as deprecated (they'll be cleaned next cycle)
        valid_ids = set(eid for eid in evicted_ids if eid)
        for m in all_memories:
            if m.get("id") in valid_ids:
                m["metadata"]["deprecated"] = "true"

        self._rewrite_file(storage, all_memories)
        return len(valid_ids)

    # ── full cycle ─────────────────────────────────────────────────────

    def run_cycle(self) -> Dict[str, Any]:
        """Run a complete decay cycle: classify, score, evict.

        Returns stats dict.
        """
        t0 = time.monotonic()
        storage = self.manager.semantic
        all_m = storage.get_all()

        # Classify all
        tier_counts = {TierLevel.HOT: 0, TierLevel.WARM: 0, TierLevel.COLD: 0}
        for m in all_m:
            tier = self.classify(m)
            tier_counts[tier] += 1

        # Evict
        evicted = self.evict()

        duration = round((time.monotonic() - t0) * 1000, 1)
        self._last_cycle = time.time()

        return {
            "total": len(all_m),
            "hot": tier_counts[TierLevel.HOT],
            "warm": tier_counts[TierLevel.WARM],
            "cold": tier_counts[TierLevel.COLD],
            "evicted": evicted,
            "duration_ms": duration,
        }

    @staticmethod
    def _rewrite_file(storage, memories: List[Dict[str, Any]]) -> None:
        """Rewrite the MEMORY.md file with updated memories."""
        with open(storage.file_path, "w", encoding="utf-8") as f:
            f.write("# MEMORY.md\n\n")
            f.write("<!-- Core Memory - Permanent Storage -->\n\n")
            for mem in memories:
                f.write(storage._format_memory(mem))

    def __repr__(self) -> str:
        return (
            f"TieredDecayEngine(hot={self.hot_ttl}s, warm={self.warm_ttl//86400}d, "
            f"cold={self.cold_ttl//86400}d)"
        )

File: src/decay/tiered_decay.ts
// Copyright 2026 Peter Cheng
// Licensed under Apache-2.0

/**
 * TieredDecayEngine (F2 - v4.7.0)
 *
 * Three-tier storage decay: HOT (current session), WARM (recently accessed),
 * COLD (long-term). Evicts based on composite score of recency, frequency,
 * and importance.
 */

import type { DecayConfig } from "./functions";

// ── External type stubs ────────────────────────────────────────────────

/**
 * Minimal interface for a storage backend that TieredDecayEngine interacts with.
 * In the full system this is SemanticStorage.
 */
export interface StorageBackend {
  filePath: string;
  getAll(): Record<string, unknown>[];
  _formatMemory(mem: Record<string, unknown>): string;
}

/**
 * Minimal interface for an LLM provider capable of text generation.
 */
export interface LLMProvider {
  generate(prompt: string, options?: { maxTokens?: number; system?: string }): string;
}

// ── Prompt ─────────────────────────────────────────────────────────────

const IMPORTANCE_PROMPT =
  "Rate the importance of this memory on a scale of 0.0 to 1.0 " +
  "(0=trivial, 0.5=useful, 0.8=very important, 1.0=critical). " +
  "Return only the number.\n\nMemory: {content}";

// ── Enum ───────────────────────────────────────────────────────────────

export enum TierLevel {
  HOT = "hot",
  WARM = "warm",
  COLD = "cold",
}

// ── Helpers ────────────────────────────────────────────────────────────

function nowIso(): string {
  return new Date().toISOString();
}

// ── Engine ─────────────────────────────────────────────────────────────

export class TieredDecayEngine {
  /** Track access timestamps for recency scoring */
  private _accessLog: Record<string, number[]> = {};
  private _importanceCache: Record<string, number> = {};
  private _lastCycle: number = 0;

  hotTtl: number;
  warmTtl: number;
  coldTtl: number;
  maxHot: number;
  maxWarm: number;
  maxCold: number;

  constructor(
    private _storage: StorageBackend,
    private _config?: DecayConfig,
    private _llmProvider?: LLMProvider,
    hotTtl: number = 3600,
    warmTtlDays: number = 7,
    coldTtlDays: number = 30,
    maxHot: number = 100,
    maxWarm: number = 500,
    maxCold: number = 2000,
  ) {
    this.hotTtl = hotTtl; // seconds
    this.warmTtl = warmTtlDays * 86400;
    this.coldTtl = coldTtlDays * 86400;
    this.maxHot = maxHot;
    this.maxWarm = maxWarm;
    this.maxCold = maxCold;
  }

  // ── classification ───────────────────────────────────────────────────

  /** Classify a single memory into a tier based on its age and metadata. */
  classify(memory: Record<string, unknown>): TierLevel {
    const meta = (memory.metadata as Record<string, string>) || {};

    // Deprecated memories go directly to COLD for immediate eviction
    if (meta.deprecated === "true" || meta.deprecated === "True" || meta.deprecated === "1") {
      return TierLevel.COLD;
    }

    // Use creation timestamp for age calculation
    const createdRaw =
      (memory.created_at as string) || (memory.timestamp as string) || nowIso();
    const createdDt = new Date(createdRaw.replace("Z", "+00:00"));
    const ageSeconds = (Date.now() - createdDt.getTime()) / 1000;

    if (ageSeconds <= this.hotTtl) return TierLevel.HOT;
    if (ageSeconds <= this.warmTtl) return TierLevel.WARM;
    return TierLevel.COLD;
  }

  /**
   * Record an access, potentially promoting the memory tier.
   *
   * Promotion rules:
   *   COLD -> WARM if recently accessed (resets TTL)
   *   WARM -> HOT if accessed within hot TTL
   *   HOT stays HOT
   */
  promote(memoryId: string): TierLevel | null {
    if (!memoryId) return null;

    const now = Date.now() / 1000;
    if (!this._accessLog[memoryId]) {
      this._accessLog[memoryId] = [];
    }
    this._accessLog[memoryId].push(now);

    // Look up current memory
    const allM = this._storage.getAll();
    const mem = allM.find((m) => (m.id as string) === memoryId);
    if (!mem) return null;

    const current = this.classify(mem);
    if (current === TierLevel.HOT || current === TierLevel.WARM) return current;

    // COLD -> WARM promotion: record access and return
    const freq = this._accessFrequency(memoryId);
    return freq >= 2 ? TierLevel.WARM : TierLevel.COLD;
  }

  // ── importance scoring ───────────────────────────────────────────────

  /** Score memory importance (0.0 - 1.0). Uses cached LLM score or rule-based fallback. */
  getImportance(memory: Record<string, unknown>): number {
    const mid = (memory.id as string) || "";
    if (mid && mid in this._importanceCache) return this._importanceCache[mid];

    // Try LLM scoring
    if (this._llmProvider) {
      try {
        const content = (memory.content as string) || "";
        const scoreText = this._llmProvider.generate(
          IMPORTANCE_PROMPT.replace("{content}", content),
          { maxTokens: 16 },
        );
        const score = this._parseScore(scoreText);
        if (score !== null) {
          if (mid) this._importanceCache[mid] = score;
          return score;
        }
      } catch {
        // fall through to rule-based
      }
    }

    // Rule-based fallback
    const score = this._ruleImportance(memory);
    if (mid) this._importanceCache[mid] = score;
    return score;
  }

  private _parseScore(text: string): number | null {
    if (!text) return null;
    const score = parseFloat(text.trim());
    if (isNaN(score)) return null;
    return Math.max(0.0, Math.min(1.0, score));
  }

  /** Rule-based importance fallback when LLM is unavailable. */
  private _ruleImportance(memory: Record<string, unknown>): number {
    const content = (memory.content as string) || "";
    const tags = (memory.tags as string[]) || [];
    const meta = (memory.metadata as Record<string, string>) || {};
    let score = 0.3; // baseline

    // Longer content may be more important
    if (content.length > 200) score += 0.15;
    else if (content.length > 50) score += 0.05;

    // Presence of tags
    if (tags.length > 0) score += 0.05;

    // Critical tags
    const criticalKeywords = ["critical", "important", "essential", "preference", "rule"];
    if (tags.some((t) => criticalKeywords.some((kw) => t.toLowerCase().includes(kw)))) {
      score += 0.2;
    }

    // Explicit importance in metadata
    if (meta.importance !== undefined) {
      const parsed = parseFloat(meta.importance);
      if (!isNaN(parsed)) score = parsed;
    }

    return Math.max(0.0, Math.min(1.0, score));
  }

  // ── access metrics ───────────────────────────────────────────────────

  private _accessFrequency(memoryId: string): number {
    const timestamps = this._accessLog[memoryId] || [];
    const now = Date.now() / 1000;
    const cutoff = now - 14 * 86400;
    return timestamps.filter((ts) => ts >= cutoff).length;
  }

  /** Seconds since last access (lower = more recent). */
  private _accessRecency(memoryId: string): number {
    const timestamps = this._accessLog[memoryId] || [];
    if (timestamps.length === 0) return Infinity;
    return Date.now() / 1000 - Math.max(...timestamps);
  }

  // ── eviction ─────────────────────────────────────────────────────────

  /**
   * Composite eviction score. Lower score = evict first.
   *
   * Formula: norm_recency * 0.4 + norm_freq * 0.3 + importance * 0.3
   */
  private _compositeScore(memory: Record<string, unknown>): number {
    const mid = (memory.id as string) || "";
    const importance = this.getImportance(memory);

    const recency = this._accessRecency(mid);
    let normRec: number;
    if (!isFinite(recency) || recency <= 0) {
      normRec = 0.0;
    } else {
      const ttlX2 = this.coldTtl * 2;
      normRec = Math.max(0.0, 1.0 - recency / ttlX2);
    }

    const freq = this._accessFrequency(mid);
    const normFreq = Math.min(1.0, freq / 5.0); // cap at 5 accesses

    return normRec * 0.4 + normFreq * 0.3 + importance * 0.3;
  }

  /**
   * Evict low-score memories, preferentially from COLD tier.
   * Always evicts deprecated memories first, then applies composite scoring.
   * Returns count of evicted memories.
   */
  evict(): number {
    const allMemories = this._storage.getAll();

    const evictedIds: string[] = [];
    const nonEvicted: Record<string, unknown>[] = [];

    // Phase 1: always evict deprecated
    for (const m of allMemories) {
      const meta = (m.metadata as Record<string, string>) || {};
      if (meta.deprecated === "true" || meta.deprecated === "True" || meta.deprecated === "1") {
        evictedIds.push((m.id as string) || "");
      } else {
        nonEvicted.push(m);
      }
    }

    // Respect max tier capacities
    const tiers: Record<TierLevel, Record<string, unknown>[]> = {
      [TierLevel.HOT]: [],
      [TierLevel.WARM]: [],
      [TierLevel.COLD]: [],
    };
    for (const m of nonEvicted) {
      const tier = this.classify(m);
      tiers[tier].push(m);
    }

    // Evict from WARM and COLD tiers if over capacity
    const maxMap: Record<TierLevel, number> = {
      [TierLevel.HOT]: this.maxHot,
      [TierLevel.WARM]: this.maxWarm,
      [TierLevel.COLD]: this.maxCold,
    };

    for (const [tier, maxCap] of Object.entries(maxMap)) {
      if (tier === TierLevel.HOT) continue; // Don't evict from HOT tier
      const tierMems = tiers[tier as TierLevel];
      const overflow = tierMems.length - maxCap;
      if (overflow <= 0) continue;

      // Score and sort: lowest score first
      const scored = tierMems.map((m) => ({ score: this._compositeScore(m), mem: m }));
      scored.sort((a, b) => a.score - b.score);
      for (const { mem } of scored.slice(0, overflow)) {
        evictedIds.push((mem.id as string) || "");
      }
    }

    if (evictedIds.length === 0) return 0;

    // Mark evicted memories as deprecated
    const validIds = new Set(evictedIds.filter(Boolean));
    for (const m of allMemories) {
      const id = m.id as string;
      if (validIds.has(id)) {
        const meta = (m.metadata as Record<string, string>) || {};
        meta.deprecated = "true";
        m.metadata = meta;
      }
    }

    this._rewriteFile(this._storage, allMemories);
    return validIds.size;
  }

  // ── full cycle ───────────────────────────────────────────────────────

  /** Run a complete decay cycle: classify, score, evict. Returns stats dict. */
  runCycle(): Record<string, unknown> {
    const t0 = Date.now();
    const allM = this._storage.getAll();

    // Classify all
    const tierCounts: Record<TierLevel, number> = {
      [TierLevel.HOT]: 0,
      [TierLevel.WARM]: 0,
      [TierLevel.COLD]: 0,
    };
    for (const m of allM) {
      const tier = this.classify(m);
      tierCounts[tier]++;
    }

    // Evict
    const evicted = this.evict();

    const duration = Math.round(Date.now() - t0);
    this._lastCycle = Date.now() / 1000;

    return {
      total: allM.length,
      hot: tierCounts[TierLevel.HOT],
      warm: tierCounts[TierLevel.WARM],
      cold: tierCounts[TierLevel.COLD],
      evicted,
      durationMs: duration,
    };
  }

  /** Rewrite the MEMORY.md file with updated memories. */
  private _rewriteFile(storage: StorageBackend, memories: Record<string, unknown>[]): void {
    const fs = require("fs") as typeof import("fs");
    const lines: string[] = [];
    lines.push("# MEMORY.md\n");
    lines.push("<!-- Core Memory - Permanent Storage -->\n");
    for (const mem of memories) {
      lines.push(storage._formatMemory(mem));
    }
    fs.writeFileSync(storage.filePath, lines.join("\n"), "utf-8");
  }

  toString(): string {
    return `TieredDecayEngine(hot=${this.hotTtl}s, warm=${this.warmTtl / 86400}d, cold=${this.coldTtl / 86400}d)`;
  }
}
